@here/harp-mapview/lib/text/Placement.ts

```diff
@@ -313,46 +345,118 @@ export function placePointLabel(
+    const placements = possibleAnchorPlacementsCW;
+    // Find first anchor placement.
+    const idx = placements.findIndex(val => {
+        return val.h === basePlacement.h && val.v === basePlacement.v;
+    });
+    // If not listed in placement options list (idx === -1) will start from current alignment setup.
+    const placementsNum = placements.length;
+    let textOffset: THREE.Vector2;
+    // Will be true if all text placements are invisible.
+    let allInvisible: boolean = true;
+    let i = 0;
+    for (i = 0; i < placementsNum; ++i) {
+        const anchorPlacement = idx + i < 0 ? basePlacement : placements[(idx + i) % placementsNum];
+        // Override label text alignment for measurements and leave it so if passed collisions test.
+        // NOTE: This actually writes to label.layoutStyle.verticalAlignment/horizontalAlignment.
+        textCanvas.textLayoutStyle.horizontalAlignment = anchorPlacement.h;
+        textCanvas.textLayoutStyle.verticalAlignment = anchorPlacement.v;
+
+        // Compute label bounds according to layout settings.
+        textCanvas.measureText(label.glyphs!, label.bounds, tmpMeasurementParams);
+        // Compute text offset from the anchor point
+        textOffset = computePointTextOffset(label, scale, tmpTextOffset);
+        textOffset.add(screenPosition);
+        tmpBox.copy(label.bounds!);
+        tmpBox.min.multiplyScalar(scale);
+        tmpBox.max.multiplyScalar(scale);
+        // Add margin after scaling, this ensures the margin is consistent across all
+        // labels - regardless of distance scaling (or any other) factor.
+        // TODO: Make the margin configurable
+        tmpBox.expandByVector(pointLabelMargin);
+        tmpBox.translate(textOffset);
+        tmp2DBox.set(
+            tmpBox.min.x,
+            tmpBox.min.y,
+            tmpBox.max.x - tmpBox.min.x,
+            tmpBox.max.y - tmpBox.min.y
+        );
 
-    if (isRejected || (!label.textMayOverlap && screenCollisions.isAllocated(tmp2DBox))) {
-        return labelState.visible ? PlacementResult.Rejected : PlacementResult.Invisible;
-    }
+        // Check the text visibility.
+        if (!screenCollisions.isVisible(tmp2DBox)) {
+            continue;
+            //return PlacementResult.Invisible;
+        } else {
+            allInvisible = false;
+        }
+
+        // If icon rejected proceed to test further if all elements are invisible.
+        // TODO:
+        // Check what the old code actually gives for callers?
+        // If icon's is isRejected, should we test visibility of all placements, maybe better to
+        // interrupt tests now and save some CPU time?
+        if (isRejected) {
+            continue;
+            //return labelState.visible ? PlacementResult.Rejected : PlacementResult.Invisible;
+        }
+        // Check label's text collision.
+        // TODO: Same comment as above, maybe better to interrupt the tests here after checking
```

Comment:
you can't interrupt the test here, this condition means the placement alternative failed, you need to test the rest.

Reply:
What I meant is that this logic is kind of awkward and I guess should be moved to caller (TextElementsRenderer). For me this function should simply return: Ok, Invisible (all invisible regardless of icon / label state), Rejected no solution found.
Currently (and in original version) there is a logic that complicates it a lot, because we are trying to make some deduction just for fading logic... This in turn results in unnecessary calculations for labels that are rejected because of icon, we would need to chat F2F to explain the details.

Reply:
Entire code has changed after function spiting.